### ISSUE TITLE
Handling of dirichlet dofs potentially worsens operator condition

### ISSUE THREAD

**Issue by andreasbuhr** [pymor/pymor#113]
Hi pymor,

for Dirichlet-dofs, an operator gets a "1" on the diagonal and the rhs gets the Dirichlet value set. This is done in operators/cg.py. However, this may worsen the condition of the operator: E.g. if all other eigenvalues are of order 1e5, adding an eigenvalue of 1e0 leads to bad behavior of sparse solvers.

It would be better to put a value on the diagonal which is of comparable size to the other entries in the matrix.

**Comment by sdrave** [comment]
@ftalbrecht: This might also be of relevance for dune-gdt?

**Comment by ftalbrecht** [comment]
I do hope that we do set the rhs to 0, instead of 1 (and apply a dirichlet correction shift, etc...). In that case we could set the diagonal entry of the lhs to any non-zero value, you are right. Do you have any idea yet how to efficiently find a good value? Do you also happen to have a minimal working example lying around? :)

**Comment by ftalbrecht** [comment]
The way we apply constraints in dune-gdt does not make it particularly easy to set the entry to any value other than 1, since we do it locally and prepare the constraints without knowing the matrix...
But since we do have other problems conecrning contraints we might refactor them if we find a nice solution here...

**Comment by sdrave** [comment]
The problem is that pyMOR clears Dirichlet rows and sets the diagonal to 1 while all other entries scale with the grid size. This has nothing to do with the rhs ..

**Comment by sdrave** [comment]
I guess for pyMOR, one could simply leave the diagonal entry as it was before boundary handling (which would correspond to the diagonal entry for a Neumann dof).

**Comment by andreasbuhr** [comment]
maybe it's not so simple, as the corresponding entry in the rhs also has to be scaled!?

**Comment by ftalbrecht** [comment]
@andreasbuhr: no, I think it does not have to be, since the entry in the rhs should be zero (see my first comment). @sdrave: letting the diagonal value simple be what it was before might be a reasonable idea (as long as it is not zero).
I might investigate in the context of the benchmarks...

**Comment by andreasbuhr** [comment]
@ftalbrecht 
The entry in the rhs should be the Dirichlet value. So for Dirichlet-zero boundary values you are right.

**Comment by ftalbrecht** [comment]
No it should not. Especially in the context of model reduction any constraints have to be applied with a zero entry in the right hand side (you can not span a reduced space with functions having nonzero boundary values in the context of CG FEM). You can transform any dirichlet problem into one with zero boundary values, solve the system, and transform back. I am not sure if Stephan implemented it this way in pyMOR, though. But I do not want to discuss this any further here since it is not really relevant. For me (as in: the dun-gdt developer) at least, it makes things easier, since I do not have to rescale the entries of the rhs.

**Comment by andreasbuhr** [comment]
@ftalbrecht :
You are right the rhs should be zero after applying the Dirichlet shift. However, for calculating the Dirichlet shift, you need the operator and the original rhs, and the original rhs needs to have the (potenetially scaled) dirichlet values. So when using just the entry which is there (instead of one) on the diagonal for the dirichlet values, you somehow have to channel this value into the rhs. 
Or do have another solution on how to calculate the Dirichlet shift?

**Comment by andreasbuhr** [comment]
OK, i figured:
If you have not 1 on the diagonal of the matrix of the operator for the Dirichlet values but some other value, say c, then you also need to scale the corresponding Dirichlet entries in the rhs by c. Then you can calculate the correct Dirichlet shift using the operator and the rhs. 
When neglecting the scaling by c in the rhs, the Dirichlet shift will be wrong. However, only the Dirichlet entries of the Dirichlet shift will be wrong. And they are known to be zero. Solution: just set them to zero. So there are two options:

1. We get the value c in the calculation of the rhs and then calculate the correct Dirichlet shift.
2. We do not use the scaling c in the calculation of the rhs, calculate the (wrong) Dirichlet shift and correct it by setting the Dirichlet values to zero.

Opinions?

**Comment by ftalbrecht** [comment]
I have considered the problem of nonzero boundary values in the context of RB in great detail and would like to postpone the question for nonzero boundary values until I have a little more time to explain mayself, or until I am back in Münster. Until then, I think it would be a good start to consider the case of zero boundary values first. If we find a good heuristic of how to deal with this problem and if it really shows a performance gain then we can come back to the case of nonzero Dirichlet values.
